tools/pubsys/src/aws/validate_ssm/results.rs

```diff
@@ -0,0 +1,681 @@
+}
+
+impl From<&HashSet<SsmValidationResult>> for SsmValidationRegionSummary {
+    fn from(results: &HashSet<SsmValidationResult>) -> Self {
+        let mut region_validation = SsmValidationRegionSummary {
+            correct: 0,
+            incorrect: 0,
+            missing: 0,
+            unexpected: 0,
+            accessible: true,
+        };
+        for validation_result in results {
+            match validation_result.status {
+                SsmValidationResultStatus::Correct => region_validation.correct += 1,
+                SsmValidationResultStatus::Incorrect => region_validation.incorrect += 1,
+                SsmValidationResultStatus::Missing => region_validation.missing += 1,
+                SsmValidationResultStatus::Unexpected => region_validation.unexpected += 1,
+            }
+        }
+        region_validation
+    }
+}
+
+impl SsmValidationRegionSummary {
+    fn no_valid_results() -> Self {
+        // When the parameters in a region couldn't be retrieved, use `-1` to indicate this in the output table
+        // and set `accessible` to `false`
+        SsmValidationRegionSummary {
+            correct: -1,
+            incorrect: -1,
+            missing: -1,
+            unexpected: -1,
+            accessible: false,
+        }
+    }
+}
+
+/// Represents all SSM validation results
+#[derive(Debug)]
+pub struct SsmValidationResults {
+    pub(crate) results: HashMap<Region, Result<HashSet<SsmValidationResult>>>,
+}
+
+impl Default for SsmValidationResults {
+    fn default() -> Self {
+        Self::new(HashMap::new())
+    }
+}
+
+impl Display for SsmValidationResults {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // Create a summary for each region, counting the number of parameters per status
+        let region_validations: HashMap<Region, SsmValidationRegionSummary> = self
+            .results
+            .iter()
+            .map(|(region, region_result)| {
+                region_result
+                    .as_ref()
+                    .map(|region_validation| {
+                        (
+                            region.clone(),
+                            SsmValidationRegionSummary::from(region_validation),
+                        )
+                    })
+                    .unwrap_or((
+                        region.clone(),
+                        SsmValidationRegionSummary::no_valid_results(),
+                    ))
+            })
+            .collect();
+
+        // Represent the HashMap of summaries as a `Table`
+        let table = Table::new(
+            region_validations
+                .iter()
+                .map(|(region, results)| (region.to_string(), results))
+                .collect::<Vec<(String, &SsmValidationRegionSummary)>>(),
+        )
+        .to_string();
+        write!(f, "{}", table)
+    }
+}
+
+impl SsmValidationResults {
+    pub fn new(results: HashMap<Region, Result<HashSet<SsmValidationResult>>>) -> Self {
```

Comment:
Seems weird. Shouldn't the Result be unwrapped somewhere else?

```suggestion
    pub fn new(results: HashMap<Region, HashSet<SsmValidationResult>>) -> Self {
```

Reply:
We keep the Result there to indicate the difference between "couldn't retrieve the SSM parameters in that region for some reason" and "some of the retrieved parameters were wrong/missing/unexpected".